### PR TITLE
fix: m_type_declaration corrected for procedure declaration

### LINKED ISSUE #5142: (PRIMA) Error while using module which contains procedure(external_subroutine)
Towards #4397 

**file 1** : 
```fortran
module temp
contains
    subroutine OBJCON()
    end subroutine OBJCON
end module temp

module evaluate_mod
contains
subroutine evaluate(calcfc)
    use temp
    implicit none
    procedure(OBJCON) :: calcfc 
end subroutine evaluate
end module evaluate_mod
```
**file 2** : 
```fortran
program initialize
use, non_intrinsic :: evaluate_mod
end program initialize
```

**LFortran Error** : 
```
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/c/lfortran/src/bin/lfortran", in _start
  File "/build/glibc-LcI20x/glibc-2.31/csu/../csu/libc-start.c", line 308, in __libc_start_main
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2669, in main
    return main_app(argc, argv);
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 2612, in main_app(int, char**)
    err = compile_src_to_object_file(arg_file, tmp_o, false,
  File "/mnt/c/lfortran/src/bin/lfortran.cpp", line 1000, in (anonymous namespace)::compile_src_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::CompilerOptions&, LCompilers::PassManager&)
    result = fe.get_asr2(input, lm, diagnostics);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 270, in LCompilers::FortranEvaluator::get_asr2(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::LocationManager&, LCompilers::diag::Diagnostics&)
    Result<ASR::TranslationUnit_t*> res2 = get_asr3(*ast, diagnostics);
  File "/mnt/c/lfortran/src/lfortran/fortran_evaluator.cpp", line 292, in LCompilers::FortranEvaluator::get_asr3(LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&)
    compiler_options.symtab_only, compiler_options);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_to_asr.cpp", line 73, in LCompilers::LFortran::ast_to_asr(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, bool, LCompilers::CompilerOptions&)
    instantiate_types, instantiate_symbols, entry_functions, entry_function_arguments_mapping, data_structure);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 3499, in LCompilers::LFortran::symbol_table_visitor(Allocator&, LCompilers::LFortran::AST::TranslationUnit_t&, LCompilers::diag::Diagnostics&, LCompilers::SymbolTable*, LCompilers::CompilerOptions&, std::map<unsigned long, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned long, LCompilers::ASR::symbol_t*, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, LCompilers::ASR::symbol_t*> > >&, std::map<unsigned long, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::ttype_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::ttype_t*> > > > > >&, std::map<unsigned int, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, LCompilers::ASR::symbol_t*, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, LCompilers::ASR::symbol_t*> > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<LCompilers::LFortran::AST::stmt_t*, std::allocator<LCompilers::LFortran::AST::stmt_t*> > > > > > > >&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<int, std::allocator<int> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<int, std::allocator<int> > > > >&, std::vector<LCompilers::ASR::stmt_t*, std::allocator<LCompilers::ASR::stmt_t*> >&)
    v.visit_TranslationUnit(ast);
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 120, in LCompilers::LFortran::SymbolTableVisitor::visit_TranslationUnit(LCompilers::LFortran::AST::TranslationUnit_t const&)
    visit_ast(*x.m_items[i]);
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4884, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_ast(LCompilers::LFortran::AST::ast_t const&)
    void visit_ast(const ast_t &b) { visit_ast_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4841, in void LCompilers::LFortran::AST::visit_ast_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::ast_t const&, LCompilers::LFortran::SymbolTableVisitor&)
    case astType::mod: { v.visit_mod((const mod_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4887, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_mod(LCompilers::LFortran::AST::mod_t const&)
    void visit_mod(const mod_t &b) { visit_mod_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4474, in void LCompilers::LFortran::AST::visit_mod_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::mod_t const&, LCompilers::LFortran::SymbolTableVisitor&)
    case modType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 520, in LCompilers::LFortran::SymbolTableVisitor::visit_Program(LCompilers::LFortran::AST::Program_t const&)
    visit_unit_decl1(*x.m_use[i]);
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4896, in LCompilers::LFortran::AST::BaseVisitor<LCompilers::LFortran::SymbolTableVisitor>::visit_unit_decl1(LCompilers::LFortran::AST::unit_decl1_t const&)
    void visit_unit_decl1(const unit_decl1_t &b) { visit_unit_decl1_t(b, self()); }
  File "/mnt/c/lfortran/src/lfortran/ast.h", line 4492, in void LCompilers::LFortran::AST::visit_unit_decl1_t<LCompilers::LFortran::SymbolTableVisitor>(LCompilers::LFortran::AST::unit_decl1_t const&, LCompilers::LFortran::SymbolTableVisitor&)
    case unit_decl1Type::Use: { v.visit_Use((const Use_t &)x); return; }
  File "/mnt/c/lfortran/src/lfortran/semantics/ast_symboltable_visitor.cpp", line 2607, in LCompilers::LFortran::SymbolTableVisitor::visit_Use(LCompilers::LFortran::AST::Use_t const&)
    t = (ASR::symbol_t*)(ASRUtils::load_module(al, tu_symtab,
  File "/mnt/c/lfortran/src/libasr/asr_utils.cpp", line 308, in LCompilers::ASRUtils::load_module(Allocator&, LCompilers::SymbolTable*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::Location const&, bool, LCompilers::PassOptions&, bool, std::function<void (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::Location const&)>)
    ASR::TranslationUnit_t *mod1 = find_and_load_module(al, module_name,
  File "/mnt/c/lfortran/src/libasr/asr_utils.cpp", line 457, in LCompilers::ASRUtils::find_and_load_module(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::SymbolTable&, bool, LCompilers::PassOptions&)
    ASR::TranslationUnit_t *asr = load_modfile(al, modfile, false, symtab);
  File "/mnt/c/lfortran/src/libasr/modfile.cpp", line 93, in LCompilers::load_modfile(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::SymbolTable&)
    ASR::asr_t *asr = deserialize_asr(al, asr_binary, load_symtab_id, symtab);
  File "/mnt/c/lfortran/src/libasr/serialization.cpp", line 414, in LCompilers::deserialize_asr(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, LCompilers::SymbolTable&)
    return deserialize_asr(al, s, load_symtab_id);
  File "/mnt/c/lfortran/src/libasr/serialization.cpp", line 420, in LCompilers::deserialize_asr(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)
    ASR::asr_t *node = v.deserialize_node();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 35906, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_node()
    case (ASR::asrType::unit) : return self().deserialize_unit();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 35952, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_unit()
    case (ASR::unitType::TranslationUnit) : return self().deserialize_TranslationUnit();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 35936, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_TranslationUnit()
    symbol_t *sym = down_cast<symbol_t>(deserialize_symbol());
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36555, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_symbol()
    case (ASR::symbolType::Module) : return self().deserialize_Module();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36010, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_Module()
    symbol_t *sym = down_cast<symbol_t>(deserialize_symbol());
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36556, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_symbol()
    case (ASR::symbolType::Function) : return self().deserialize_Function();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36044, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_Function()
    symbol_t *sym = down_cast<symbol_t>(deserialize_symbol());
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36563, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_symbol()
    case (ASR::symbolType::Variable) : return self().deserialize_Variable();
  File "/mnt/c/lfortran/src/libasr/../libasr/asr.h", line 36354, in LCompilers::ASR::DeserializationBaseVisitor<LCompilers::ASRDeserializationVisitor>::deserialize_Variable()
    m_type_declaration = self().read_symbol();
AssertFailed: id_symtab_map.find(symtab_id) != id_symtab_map.end()
```

It works with GFortran.

**Some things I noticed with LFortran**: 
1) If we keep both code in one file it works.
2) If we remove Procedure(OBJCON) then it works. 
3) If I keep OBJCON subroutine in evalute_mod module itself then it works.

### PR DESCRIPTION
**Pull request by jinangshah21**
Fixes #5142 
Here I didn't add test because this issue needs 2 separated files.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3631,10 +3631,14 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
-            type = ASR::down_cast<ASR::Function_t>(v)->m_function_signature;
-            type_declaration = v;
+            ASR::symbol_t *temp = ASRUtils::symbol_get_past_external(v);
+            LCOMPILERS_ASSERT(ASR::is_a<ASR::Function_t>(*temp));
+            type = ASR::down_cast<ASR::Function_t>(temp)->m_function_signature;
+            if(ASR::is_a<ASR::ExternalSymbol_t>(*v)) {
+                type_declaration = v;
+            } else {
+                type_declaration = temp;
+            }
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:3641]: The logic doesn't seem to make sense: if `v` is ExternalSymbol, then `v` is used. If it is not an external symbol, then `temp` should be equal to `v`, right?

Comment (jinangshah21) [reply src/lfortran/semantics/ast_common_visitor.h]: Corrected it.

**Comment by certik** [comment]
Add a test.

**Comment by certik** [comment]
Just add both files. See "EXTRAFILES" in `integration_tests/CMakeLists.txt`.

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
Add two new `.f90` files, supposedly `test_01.f90` and `test_01_b.f90` and then register it as:

```CMake
RUN(NAME test_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc EXTRAFILES test_01_b.f90)
```

**Review by Pranavchiku — CHANGES_REQUESTED** [review]
You also need to apply the following diff in `asr_verify.cpp` to keep things in sync:

```diff
diff --git a/src/libasr/asr_verify.cpp b/src/libasr/asr_verify.cpp
index ebe25879e..5da47d58b 100644
--- a/src/libasr/asr_verify.cpp
+++ b/src/libasr/asr_verify.cpp
@@ -930,7 +930,7 @@ public:
             ASR::symbol_t *s = ASRUtils::symbol_get_past_external(x.m_name);
             if (ASR::is_a<ASR::Variable_t>(*s)) {
                 ASR::Variable_t *v = ASR::down_cast<ASR::Variable_t>(s);
-                require(v->m_type_declaration && ASR::is_a<ASR::Function_t>(*v->m_type_declaration),
+                require(v->m_type_declaration && ASR::is_a<ASR::Function_t>(*ASRUtils::symbol_get_past_external(v->m_type_declaration)),
                     "SubroutineCall::m_name '" + std::string(symbol_name(x.m_name)) + "' is a Variable, but does not point to Function");
                 require(ASR::is_a<ASR::FunctionType_t>(*v->m_type),
                     "SubroutineCall::m_name '" + std::string(symbol_name(x.m_name)) + "' is a Variable, but the type is not FunctionType");
```

File: src/libasr/asr_verify.cpp [hunk 2]
```diff
@@ -930,7 +930,7 @@ class VerifyVisitor : public BaseWalkVisitor<VerifyVisitor>
             ASR::symbol_t *s = ASRUtils::symbol_get_past_external(x.m_name);
             if (ASR::is_a<ASR::Variable_t>(*s)) {
                 ASR::Variable_t *v = ASR::down_cast<ASR::Variable_t>(s);
-                require(v->m_type_declaration && ASR::is_a<ASR::Function_t>(*v->m_type_declaration),
+                require(v->m_type_declaration && ASR::is_a<ASR::Function_t>(*ASRUtils::symbol_get_past_external(v->m_type_declaration)),
```
Comment (certik) [inline src/libasr/asr_verify.cpp:933]: I don't see any `SubroutineCall` nodes in the above test case. Is this being tested? We should add a test for this.

**Review by certik — APPROVED** [review]
I think it's fine, but I want to have tests for all new changes, and the ASR verify change doesn't seem to be tested (in other words, doesn't seem to be needed).

**Comment by certik** [comment]
I merged it. If the ASR change is not needed, we can remove it in a subsequent PR, or add a test.

**Comment by Pranavchiku** [comment]
We need it, PRIMA compilation fails without it. @jinangshah21 you can revert asr_verify.cpp change locally and compile PRIMA, it will throw Asr Verify error and you can create MRE. The fix for MRE will be the diff above.

**Comment by jinangshah21** [comment]
> I think it's fine, but I want to have tests for all new changes, and the ASR verify change doesn't seem to be tested (in other words, doesn't seem to be needed).

It is needed because when we will call calcfc it will visit SubroutineCall :  
```fortran 
module eval_dependency
contains
    subroutine OBJCON()
    end subroutine OBJCON
end module eval_dependency

module procedure_decl_01_b
contains
subroutine evaluate(calcfc)
    use eval_dependency
    implicit none
    procedure(OBJCON) :: calcfc 
    **call calcfc()**
end subroutine evaluate
end module procedure_decl_01_b
```

**Comment by certik** [comment]
Perfect. Please send a PR adding `call calcfc()` to the test, you forgot to add it.

**Pull request merged**